src/main/java/org/folio/circulation/resources/handlers/LoanRelatedFeeFineClosedHandlerResource.java

```diff
@@ -0,0 +1,115 @@
+import static org.folio.circulation.support.ValidationErrorFailure.singleValidationError;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.folio.circulation.StoreLoanAndItem;
+import org.folio.circulation.domain.Account;
+import org.folio.circulation.domain.AccountRepository;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.policy.LostItemPolicyRepository;
+import org.folio.circulation.domain.subscribers.LoanRelatedFeeFineClosedEvent;
+import org.folio.circulation.resources.Resource;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.RouteRegistration;
+import org.folio.circulation.support.http.server.NoContentResponse;
+import org.folio.circulation.support.http.server.ValidationError;
+import org.folio.circulation.support.http.server.WebContext;
+
+import io.vertx.core.http.HttpClient;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class LoanRelatedFeeFineClosedHandlerResource extends Resource {
+
+  public LoanRelatedFeeFineClosedHandlerResource(HttpClient client) {
+    super(client);
+  }
+
+  @Override
+  public void register(Router router) {
+    new RouteRegistration("/circulation/handlers/loan-related-fee-fina-closed", router)
+      .create(this::handleFeeFineClosedEvent);
+  }
+
+  private void handleFeeFineClosedEvent(RoutingContext routingContext) {
+    final WebContext context = new WebContext(routingContext);
+
+    createAndValidateRequest(routingContext)
+      .after(request -> processEvent(context, request))
+      .thenApply(r -> r.toFixedValue(NoContentResponse::noContent))
+      .thenAccept(context::writeResultToHttpResponse);
+  }
+
+  private CompletableFuture<Result<Loan>> processEvent(
+    WebContext context, LoanRelatedFeeFineClosedEvent event) {
+
+    final Clients clients = create(context, client);
+    final LoanRepository loanRepository = new LoanRepository(clients);
+
+    return loanRepository.getById(event.getLoanId())
+      .thenCompose(r -> r.after(loan -> {
+        if (loan.isDeclaredLost()) {
+          return closeDeclaredLostLoanIfLostFeesResolved(clients, loan);
+        }
+
+        return completedFuture(succeeded(loan));
+      }));
+  }
+
+  private CompletableFuture<Result<Loan>> closeDeclaredLostLoanIfLostFeesResolved(
+    Clients clients, Loan loan) {
+
+    final AccountRepository accountRepository = new AccountRepository(clients);
+    final LostItemPolicyRepository lostItemPolicyRepository = new LostItemPolicyRepository(clients);
+
+    return accountRepository.findAccountsForLoan(loan)
+      .thenComposeAsync(lostItemPolicyRepository::findLostItemPolicyForLoan)
+      .thenCompose(loanResult -> closeLoanAndUpdateIfNeeded(loanResult, clients));
+  }
+
+  public CompletableFuture<Result<Loan>> closeLoanAndUpdateIfNeeded(
```

Comment:
What is being updated if needed, the item?

Reply:
Loan with new action and status and the item.

Reply:
Ok, maybe the item part could be made more explicit. I don't think the `if needed` suffix is necessary, as many methods might choose not to do something.

```suggestion
  public CompletableFuture<Result<Loan>> closeLoanAndUpdateItem(
```